Framework agnostic validation

antico5 opened the pull request:
This refactor aims to restructure the LSP into a framework-agnostic design. There are some basic abstractions in place to achieve this:

### Indexer

In charge of creating project instances

```
index(folder: WorkspaceFolder): Project[]
```

Scans a folder and returns a list of project instances found.

### Project

Abstraction for each framework provider.

```
fileBelongs(uri): bool
```

Given a file uri (contract), returns true/false. `true` means that this project is interested in compiling the contract.

```
resolveImportPath(from, importPath): string | undefined
```

Resolve an import path (i.e. `hardhat/console.sol') into an absolute path

```
buildCompilation(souceUri, openDocuments): solc input + version
```

Given a specific file and a list of open editor documents, return the compiler input that can be fed into solc. Also return the version that should be used to compile it.

```
onWatchedFilesChanged(changes)
```

Projects are notified when watched files changed so they can react in any way (i.e. invalidating cache, reloading config, etc).

## General changes

### Indexing process

This process happens when the extension is initialized or after new workspace folders are added. The main steps are the following:

- Scan for projects, calling all registered Indexers (one per framework implementation).
- Perform any necessary initialization, calling `project.initialize()` on all projects, in parallel.
- Scan for all solidity files in folders.
- Associate solidity files to projects
  - For each file, all registered projects are asked if the file belongs to them. More than one may match. In this case, pick the one with highest priority
  - If there are no matches, Projectless project is assigned
- Analyze all solidity files
  - Initially this step was removed to speed up initialization. But the problem is that some actions like "find all references" wouldn't work properly. Some optimizations can still be performed in future work.

### Validation process

- Only available after indexing stage was finished. By that point, all projects are fully operational.
- Generate a request id and associate it with the file under validation
  - This solves the issue of parallel validation requests and race conditions. We care only about the last one.
- Call `buildCompilation` on the file's project (framework provider layer)
- Call `compile` on CompilationService (LSP layer). This uses the hardhat module bundled with the extension, independently of the project's type.
  - Download the solc compiler
  - Execute the solc build
- Transform solc output into validation results (i.e. diagnostics, status items)
- Send results back to the client

### Error Handling

#### Solc output errors

Means that input was built successfully but there are solidity errors handled by solc. Publish diagnostics in this case.

#### Framework build input error

Means that the project (framework) was not able to build an input for solc. The abstraction in place is an error of type BuildInputError with the following data:

- projectWideErrors: there was a problem at the project level, not specific to a file. i.e. some dependency not installed, wrong config, etc.
  - will show a status item in the UI
- fileSpecificErrors: errors that can be attributed to a file. i.e. invalid import.
  - will show diagnostics + status item on the UI

#### Other errors

Any other errors are handled by showing a status item with the error message on the UI

### Things removed from the LSP layer

- Everything hardhat-specific. Except in CompilerService which uses bundled hardhat to download and execute solc.
- `ISolProject` interface. the `Project` abstract class mostly replaced it.
- Worker processes,since they are behind the hardhat adapter now
- Some validation error specific types

### Hardhat provider implementation

#### worker processes

- A child process is spawned when the project is initialized and restarted when the config file is updated
- Most services called on the project are delegated to the worker process, since that's where the local HRE is loaded
- There's a protocol that defines messages for parent-child communication using IPC. There are request ids generated to match request/responses sent between them.
- There's a timeout (5 seconds default) that will ensure that promises on the LSP process resolve/reject even if the worker process doesn't respond requests in time.

#### fileBelongs check

If hre is loaded successfully, then to check if a contract belongs to this project, it checks that it's inside the configured `sources` folder. But it can happen that hre can't be loaded i.e. because node modules are not installed or wrong config. In this case, the project will "absorb" all files inside the project root and claim it to be theirs. This is to avoid having hardhat contracts being validated with the projectless provider (and getting confusing errors like "hardhat/console.sol" doesnt exist).

### Foundry provider implementation

- Quick initial approach to this provider. ~1 hour spent on it
- Just implements remappings in JS (reading remappings.txt file). No forge invocation yet
- buildCompilation implementation is basically projectless + remappings
- hardcoded `src` folder for contracts

kanej commented on package.json:
```diff
@@ -39,7 +39,7 @@
   "activationEvents": [
     "onLanguage:solidity",
-    "workspaceContains:**/hardhat.config.ts"
+    "workspaceContains:**/hardhat.config.{ts,js}"
```
Ha, good catch.

Do we need to add foundry.toml in here as well?

antico5 replied:
Indeed! will add it

kanej commented on client/src/setup/setupLanguageServerHooks.ts:
```diff
@@ -46,7 +46,9 @@ const startLanguageServer = (extensionState: ExtensionState): void => {
         workspace.createFileSystemWatcher("**/hardhat.config.{ts,js}"),
-        workspace.createFileSystemWatcher("**/contracts/**/*.sol"),
+        workspace.createFileSystemWatcher("**/foundry.toml"),
+        workspace.createFileSystemWatcher("**/remappings.txt"),
```
What event hooks in on changes to "remappings.txt"?

kanej replied:
Right. That makes sense.

antico5 commented:
@kanej Here's an update on what was added on the new commit (named `common`, will be rebased).
- reworked "hardhat status item" into "project status item". Each contract will show a status item with the project type and link to config file. Some amount of code was removed, including the GetSolFileDetails client-server interaction because it's not needed anymore
- fixed packaging script to use a different file for the worker process
- fixed issue with compilation error after rename

Also opened a PR with the isolated changes required for the "async import resolution". Feel free to take a look but that one can be left for later: https://github.com/NomicFoundation/hardhat-vscode/pull/276

kanej commented:
> @kanej Here's an update on what was added on the new commit (named `common`, will be rebased).
> 
>     * reworked "hardhat status item" into "project status item". Each contract will show a status item with the project type and link to config file. Some amount of code was removed, including the GetSolFileDetails client-server interaction because it's not needed anymore
> 
>     * fixed packaging script to use a different file for the worker process
> 
>     * fixed issue with compilation error after rename
> 
> 
> Also opened a PR with the isolated changes required for the "async import resolution". Feel free to take a look but that one can be left for later: #276

I am still seeing the `compilation error after rename` with the latest `common` commit.

kanej commented:
Lets quieten down the log:

![image](https://user-images.githubusercontent.com/24030/200300397-52c0b9f8-b008-412d-a5bf-8cb87f3e0424.png)

You are only going to look in here when there is an issue, so if we can keep the signal to noise high that helps us in providing support.

antico5 commented:
> > @kanej Here's an update on what was added on the new commit (named `common`, will be rebased).
> > ```
> > * reworked "hardhat status item" into "project status item". Each contract will show a status item with the project type and link to config file. Some amount of code was removed, including the GetSolFileDetails client-server interaction because it's not needed anymore
> > 
> > * fixed packaging script to use a different file for the worker process
> > 
> > * fixed issue with compilation error after rename
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also opened a PR with the isolated changes required for the "async import resolution". Feel free to take a look but that one can be left for later: #276
> 
> I am still seeing the `compilation error after rename` with the latest `common` commit.
> 
>     * note: don't see it on linux(docker)

This is super annoying, but looks like the `documents` manager does not behave the same way in linux as in windows. And that's what we are using for getting the `openDocuments`. On windows, after the rename action, the collection doesn't include the files besides just the one open, then the strategy of replacing the cache with current file + open documents doesn't work. I'll go ahead and implement a function for invalidating build caches like we had before.

antico5 commented:
> Lets quieten down the log:
> 
> ![image](https://user-images.githubusercontent.com/24030/200300397-52c0b9f8-b008-412d-a5bf-8cb87f3e0424.png)
> 
> You are only going to look in here when there is an issue, so if we can keep the signal to noise high that helps us in providing support.

Should I make these validation-related log lines `trace` level, or keep them `info` but more concise ?

kanej commented on server/src/frameworks/Hardhat/HardhatProject.ts:
```diff
@@ -0,0 +1,298 @@
+        this.logger.info(`Config file changed. Restarting worker process.`);
+
+        // Kill existing worker process
+        this.workerProcess?.kill("SIGKILL");
```
Should we update the state here to stop any messages coming in during initialization?

antico5 replied:
It happens naturally since the worker status is set to `initializing`. If some validation request happens to come at this point, it will fail with a descriptive error message

kanej commented on server/src/services/initialization/onInitialized.ts:
```diff
@@ -33,45 +29,16 @@ export const onInitialized = (
-        serverState.workspaceFolders
+        serverState.workspaceFoldersToIndex
       );
+      serverState.indexingFinished = true;
```
Is this check needed? It is used as a guard against premature validation and other actions right? Should we move the indexing to the initialization, and only return from initialization when all the projects have initialized?

kanej replied:
Lets add a detailed linear task for this.

kanej commented on server/src/services/initialization/updateAvailableSolcVersions.ts:
```diff
@@ -0,0 +1,112 @@
+
+    return _.map(data.builds, "version");
+  } catch (error) {
+    return [];
```
Maybe return the defaults?

kanej replied:
We should capture this failure in sentry as well.

kanej commented on server/src/services/rename/onRename.ts:
```diff
@@ -25,19 +21,7 @@ export const onRename = (serverState: ServerState) => {
-      // are recorded, preprocessing won't recognise that
-      // the cache is no longer valid, hence we clear it
-      // before returning
-      await invalidateWorkerPreprocessCache(
```
What is the thinking on removing this, is the case no longer needed or covered in another way?

antico5 replied:
It was no longer needed because the strategy was switched to "replace sources with openDocuments", which works nicely but yesterday I found out it doesn't work on Windows. So I brought back the invalidate approach with the `project.invalidateBuildCache()` call

kanej commented on server/src/utils/getOrInitialiseSolFileEntry.ts:
```diff
@@ -10,19 +12,13 @@ import { SolFileEntry } from "../parser/analyzer/SolFileEntry";
 
   if (solFileEntry === undefined) {
-    const project = findProjectFor({ projects }, uri);
+    const project = new ProjectlessProject(serverState, path.dirname(uri));
```
Might want to check that use of `path`.

antico5 replied:
I double checked but couldn't find any problem with it. The attribute `basePath` is stored in the os-specific path format and operations with it are performed using the `path` module. Should I test a specific scenario where it could fall short ?

kanej replied:
Cool. If we are happy our integration tests cover this across OSes, no change needed.

kanej commented on server/src/utils/getOrInitialiseSolFileEntry.ts:
```diff
@@ -10,19 +12,13 @@ import { SolFileEntry } from "../parser/analyzer/SolFileEntry";
  * Uri needs to be decoded and without the "file://" prefix.
  */
 export function getOrInitialiseSolFileEntry(
-  {
```
You can always pass the serverState to everything as a giant global, but it does mean it is more work to figure out what state a function is dependent on. If you try and test the function, then suddenly you need an entire `serverState` object or you need to lie and hack the type system.

kanej replied:
Don't feel the need to change this back. Just explaining why I seem to be doing odd things.

antico5 replied:
I agree, I didn't pay too much attention to that but it makes complete sense. In this case, it's needed because of the ProjectlessProject creation inside `getOrInitialiseSolFileEntry`, which I'd like to remove. It's there for cases where for example, you have a workspace folder open but open a random solidity file from another folder into the current workspace. Then it needs to initialize a SolFileEntry for it and its dependencies. I'll see if I can remove it

antico5 commented on server/src/frameworks/Foundry/FoundryIndexer.ts:
```diff
@@ -0,0 +1,24 @@
+      ["**/lib/**"]
+    );
+
+    return Promise.all(
```
I have no idea why it was there :sweat_smile:  will remove it

antico5 commented on server/src/frameworks/Foundry/FoundryProject.ts:
```diff
@@ -0,0 +1,286 @@
+
+      switch (error.code) {
+        case 127:
+          this.initializeError =
```
This has been my reasoning but I'm open to hear an alternative approach: I think it needs to be stored because initialization and validation are different invocations. On each validation, if the initialization was successful, a failure message is returned (and then a status item in the UI). I think the "interpretation" of the execution error (ExecError => string error reason) is ok to be in this place since the caller doesn't have to know what's going under the hood (invoking forge through shell).

antico5 commented on server/src/frameworks/Foundry/FoundryProject.ts:
```diff
@@ -0,0 +1,286 @@
+      this.configSolcVersion = config.solc || undefined; // may come as null otherwise
+
+      const rawRemappings = await runCmd(
+        "~/.foundry/bin/forge remappings",
```
Very recently I changed `forge` to `~/.foundry/bin/forge` to have it work on mac. It seems to have worked because the CI is able to run the foundry integration test now. This is on the server side only. After this comment I noticed we are also using forge on the client side for formatting. Should I also modify those client side calls?

kanej replied:
Leave the client-side for the moment. Go with your suggestion of testing for `forge` and falling back to the known locations of `forge` on different platforms. To workaround people getting different results depending on how they started vscode:

https://code.visualstudio.com/docs/terminal/advanced#_process-environment

antico5 replied:
After diving into this for some minutes I think for now it's better to keep the current approach of using the default installation paths for both linux and windows. Trying to determine if `forge` as a global is available introduces too much complexity because you need a strategy for unix as well as windows for doing that. There are libraries (i.e. `which`) that try to achieve that but after a bit of testing it doesnt behave exactly as the `which` unix command. I'd leave it simple for now and see if any users have trouble with it. From foundry docs, it seems that the installation directory is deterministic

kanej replied:
Okay, lets see how far we get with platform specific known locations.

antico5 commented on server/src/frameworks/Foundry/FoundryProject.ts:
```diff
@@ -0,0 +1,286 @@
+      });
+
+      return toUnixStyle(fs.realpathSync(resolvedPath));
+    } catch (error) {
```
The `resolveImportPath` function is designed to return either a resolved path or undefined. In the implementation, we perform a try/catch because `require.resolve` throws an error when it fails to resolve. I don't think it makes sense to report these kind of errors because it happens on each invalid import. The current analysis strategy involves just ignoring the imports that can't be crawled because they can't be resolved. From the validation standpoint, all invalid imports are shown as diagnostics.

kanej replied:
If a file imports a path `/some/path.sol`, and the resolveImportPath succeeds, but the analysis of that file `/some/path.sol` fails in the recursive analyser, does the starting file also fail analysis?

antico5 replied:
No, in that case everything work as expected, just that the invalid import (from /some/path.sol) is not crawled. But from the initial file to /some/path.sol everything is tied correctly

kanej replied:
Hmmm, I prefer that we do nothing rather than doing the wrong thing.

If we do a rename, but analysis from the file failed on an import, the rename could be wrong. Better to just fail the rename because the analysis failed.

What do you think?

kanej replied:
That is a fair point. Lets not add "fix everything" to this PR (I'll add "fix everything" to the backlog).

antico5 commented on server/src/frameworks/Hardhat/HardhatProject.ts:
```diff
@@ -0,0 +1,298 @@
+      this.workerStatus = WorkerStatus.INITIALIZING;
+
+      this.workerProcess.on("message", async (message: Message) => {
+        await this._handleMessage(message);
```
Good point. If there are any unhandled errors, it'll log them on the console automatically. But I'll add a try/catch to avoid the `async function without a catch block` part.

kanej replied:
Consider sentry as well.

A deluge of useless noise, we want to avoid; but if something is going wrong in the gap that might be something we should pay attention to.

antico5 commented on server/src/frameworks/Hardhat/HardhatProject.ts:
```diff
@@ -0,0 +1,298 @@
+
+    return Promise.race([
+      workerPromise,
+      this._requestTimeout("buildCompilation"),
```
I don't think it's needed to reject the worker promise, since the parent promise (the race one) encapsulates it. If a timeout happens, the `buildCompilation` function is rejected with the timeout error. Eventually it gets shown in the UI in a status item (default error behavior).

antico5 commented on server/src/frameworks/Hardhat/HardhatProject.ts:
```diff
@@ -0,0 +1,298 @@
+        // HRE was loaded successfully. Delegate to the worker that will use the configured sources path
+        const requestId = this._prepareRequest(resolve, reject);
+
+        this.workerProcess!.send(new FileBelongsRequest(requestId, sourceURI));
```
The function `fileBelongs` can return true, false, or throw an error (handled by caller). In the case there's a communication error (the response never comes back from the worker), a timeout error will happen. If an error happens on the worker side, an `ErrorResponseMessage` will come back from the worker with some reason. If `send` fails in the first place, the promise is rejected naturally. Maybe I'm missing a case or not understanding completely, please let me know!

antico5 commented on server/src/frameworks/Hardhat/worker/WorkerProtocol.ts:
```diff
@@ -0,0 +1,104 @@
+  INVALIDATE_BUILD_CACHE,
+}
+
+export abstract class Message {
```
Initially I thought these would have more methods, but as of now it's at least "bundling" together both the type and the factory (constructor). From hierarchy what's enforced is that some of the objects have specific properties (i.e. all must have `type`, some must have a `requestId`).

antico5 commented on server/src/frameworks/Projectless/ProjectlessIndexer.ts:
```diff
@@ -0,0 +1,9 @@
+import ProjectIndexer from "../base/ProjectIndexer";
+import ProjectlessProject from "./ProjectlessProject";
+
+export default class ProjectlessIndexer extends ProjectIndexer {
```
Ok, will convert it to a regular export

antico5 commented on server/src/frameworks/base/ProjectIndexer.ts:
```diff
@@ -0,0 +1,13 @@
+    public fileRetriever: WorkspaceFileRetriever
+  ) {}
+
+  public abstract index(folder: WorkspaceFolder): Promise<Project[]>;
```
I think there's some value in having the hierarchy in place. On one hand, the explicit connection is there (i.e. a HardhatIndexer is a ProjectIndexer so it must satify the interface), and the inheritance is handy. For example neither on `HardhatIndexer` or `FoundryIndexer` I had to add a constructor definition. What would you suggest as a refactor?

antico5 commented on server/src/frameworks/Projectless/ProjectlessProject.ts:
```diff
@@ -0,0 +1,175 @@
+    return;
+  }
+
+  private async _crawlDependencies(
```
Yes indeed. The foundry implementation was a bit rushed and shares a lot with the projectless implementation. I'll pull out some shared logic including `crawlDependencies`

kanej commented on test/integration/index.ts:
```diff
@@ -12,7 +13,15 @@ export function run(): Promise<void> {
+            .getExtension("nomicfoundation.hardhat-solidity")
+            ?.exports.isReady() !== true
+        ) {
+          await sleep(100);
```
Oooohhhhh. That is far better.

antico5 commented on server/test/services/documents/onDidChangeWatchedFiles.ts:
```diff
@@ -1,218 +1,15 @@
-      } as any;
-    });
-
-    describe("that is untracked", () => {
```
I tried to not touch it, but looking through references on this attribute, I'm not sure why is it important. I'm tempted to remove it completely, but perhaps you know of the reason why it exists?

kanej replied:
It was added in so we could answer `getOpenDocumentsInProject`, when passing those for validation.

This is instead of querying the client to get the currently open editor documents.

antico5 replied:
Probably we don't need that. On that function, it's using the `serverState.documents` collection, which has the open editor documents already. In any case, I wouldn't change it now since it's out of scope for this PR.

kanej replied:
Is my mental model correct, is `serverState.documents` all docs that have ever been opened in the editor?

My vague recollection was `getopenDocumentsInProject` was trying to limit what was passed over to the validation thread, by filtering to the currently open docs.

But that was really just an attempt to mirror the original implementation while removing the need to have a call to the front-end to and back to get open docs (which tended to timeout and fail).

Do we have a feel for how big the `serverState.documents` collection might get under heavy use in a long editor session?

antico5 replied:
Just tested to be sure. `serverState.documents` only retains the current open documents in the editor. So I feel safe removing the `tracked` flag from the SolFileEntry type.

kanej replied:
Brilliant.

antico5 commented on server/src/frameworks/Foundry/FoundryProject.ts:
```diff
@@ -0,0 +1,286 @@
+      const config = JSON.parse(
+        await runCmd("~/.foundry/bin/forge config --json", this.basePath)
+      );
+      this.sourcesPath = path.join(this.basePath, config.src);
```
Just finished testing on a foundry setup on windows and it works. Tested changing the default src folder to something else and it works as expected

kanej commented on server/src/utils/Logger.ts:
```diff
@@ -93,7 +93,7 @@ export class ConnectionLogger implements Logger {
     description: string,
     callback: () => Promise<unknown>
   ) {
-    this.info(`${description}: Start`);
```
Still seeing log output for compilation:

![image](https://user-images.githubusercontent.com/24030/201331170-18be530c-bfb8-4ad3-a043-a27a6a01492c.png)

I think these are coming from `logger.trackTime`

antico5 replied:
:man_facepalming: changed the "start" line to trace and missed the "end" one. Sorry!

kanej reviewed (approved):
I closed off some conversations I think have been addressed in subsequent PRs.

After our discussion yesterday on testing approaches (+ the irrelevance of the tracking tests), I think we are good here.
